### PR TITLE
feat: add pkce-openid-backend

### PR DESCRIPTION
**Pull request by johanv26**
## Description
This adds a generic backend based on ConfigurableOpenIdConnectAuth but
with PKCE.
This backend is inspired by the social-core way to implement PKCE.
There is a current PR in work, but for the moment, that class is not merged and accessible.
So after that is finished this has its code for `code_challenge` and `code_challenge_method`implementation.
PR: https://github.com/python-social-auth/social-core/pull/856
## Testing instructions

Please provide detailed step-by-step instructions for testing this change.

## Additional information

**Jira story**
https://edunext.atlassian.net/browse/FUTUREX-606
## Checklist for Merge

- [ ] Tested in a remote environment
- [ ] Updated documentation
- [ ] Rebased master/main
- [ ] Squashed commits

### CODE REVIEW

**Comment by johanv26** [comment]
omar-nelc could you take a look 😬

File: eox_core/social_tpa_backends.py [hunk 1]
```diff
@@ -199,3 +201,72 @@ def get_user_id(self, *args, **kwargs):
                 LOG.info("Updating uid: %s to %s", uid, slug_uid)
 
         return slug_uid
+
```
Comment (omar-nelc) [inline eox_core/social_tpa_backends.py:204]:
```suggestion
# TODO: Use the `from social_core.backends.oauth import BaseOAuth2PKCE` base class once the pull request is merged: https://github.com/python-social-auth/social-core/pull/856/files#diff-d44db201b48f2ec7cab2a0c981213a2991630567778cc6608d03fa0e3804e466R467
class BaseOAuth2PKCEMixin:
    """
    Base class for providers using OAuth2 with Proof Key for Code Exchange (PKCE).
    OAuth2 details at:
        https://datatracker.ietf.org/doc/html/rfc6749
    PKCE details at:
        https://datatracker.ietf.org/doc/html/rfc7636
    """

    PKCE_DEFAULT_CODE_CHALLENGE_METHOD = "s256"
    PKCE_DEFAULT_CODE_VERIFIER_LENGTH = 32
    USE_PKCE = True

    def create_code_verifier(self):
        name = self.name + "_code_verifier"
        code_verifier_len = self.setting(
            "PKCE_CODE_VERIFIER_LENGTH", default=self.PKCE_DEFAULT_CODE_VERIFIER_LENGTH
        )
        code_verifier = self.strategy.random_string(code_verifier_len)
        self.strategy.session_set(name, code_verifier)
        return code_verifier

    def get_code_verifier(self):
        name = self.name + "_code_verifier"
        code_verifier = self.strategy.session_get(name)
        return code_verifier

    def generate_code_challenge(self, code_verifier, challenge_method):
        method = challenge_method.lower()
        if method == "s256":
            hashed = hashlib.sha256(code_verifier.encode()).digest()
            encoded = base64.urlsafe_b64encode(hashed)
            code_challenge = encoded.decode().replace("=", "")  # remove padding
            return code_challenge
        elif method == "plain":
            return code_verifier
        else:
            raise AuthException("Unsupported code challenge method.")

    def auth_params(self, state=None):
        params = super().auth_params(state=state)

        if self.USE_PKCE:
            code_challenge_method = self.setting(
                "PKCE_CODE_CHALLENGE_METHOD",
                default=self.PKCE_DEFAULT_CODE_CHALLENGE_METHOD,
            )
            code_verifier = self.create_code_verifier()
            code_challenge = self.generate_code_challenge(
                code_verifier, code_challenge_method
            )
            params["code_challenge_method"] = code_challenge_method
            params["code_challenge"] = code_challenge
        return params

    def auth_complete_params(self, state=None):
        params = super().auth_complete_params(state=state)

        if self.USE_PKCE:
            code_verifier = self.get_code_verifier()
            params["code_verifier"] = code_verifier

        return params
```

File: eox_core/social_tpa_backends.py [hunk 2]
```diff
@@ -199,3 +201,72 @@ def get_user_id(self, *args, **kwargs):
...
+
+class ConfigurableOpenIdConnectAuthPKCE(ConfigurableOpenIdConnectAuth):
```
Comment (omar-nelc) [inline eox_core/social_tpa_backends.py:206]:
```suggestion
class ConfigurableOpenIdConnectAuthPKCE(BaseOAuth2PKCEMixin, ConfigurableOpenIdConnectAuth):
```

**Review by omar-nelc — APPROVED** [review]
Thanks @johanv26. I have added a suggestion to refactor the `BaseOAuth2PKCEMixin` and use it so we can refactor later and use the `BaseOAuth2PKCE` base class once it lands on upstream.

This usually helps with the release upgrade.

File: eox_core/social_tpa_backends.py [hunk 3]
```diff
@@ -199,3 +201,86 @@ def get_user_id(self, *args, **kwargs):
...
+    There is a current PR in working, but for the moment, that class is not merged and accesible.
+    So after that is finished we use `BaseOAuth2PKCEMixin` for `code_challenge` and `code_challenge_method`implementation.
+    PR: https://github.com/python-social-auth/social-core/pull/856
+    Block code:  https://github.com/python-social-auth/social-core/pull/856/files#diff-d44db201b48f2ec7cab2a0c981213a2991630567778cc6608d03fa0e3804e466R467-R530
+
+    """
+    name = 'config-based-openidconnect-PKCE'
+    PKCE_DEFAULT_CODE_CHALLENGE_METHOD = "s256"
+    PKCE_DEFAULT_CODE_VERIFIER_LENGTH = 32
+    USE_PKCE = True
```
Comment (omar-nelc) [inline eox_core/social_tpa_backends.py:286]: Thanks @johanv26!! Looks much better now!

Those aren't needed, because they're inherited from the mixin.

File: eox_core/social_tpa_backends.py [hunk 4]
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
...
+    PKCE details at:
+        https://datatracker.ietf.org/doc/html/rfc7636
+    """
+
+    PKCE_DEFAULT_CODE_CHALLENGE_METHOD = "s256"
+    PKCE_DEFAULT_CODE_VERIFIER_LENGTH = 32
+    USE_PKCE = True
+
+    def create_code_verifier(self):
+        name = self.name + "_code_verifier"
```
Comment (andrey-canon) [inline eox_core/social_tpa_backends.py:221]:
```suggestion
        name = f"{self.name}_code_verifier"
```

File: eox_core/social_tpa_backends.py [hunk 5]
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
...
+        code_verifier_len = self.setting(
+            "PKCE_CODE_VERIFIER_LENGTH", default=self.PKCE_DEFAULT_CODE_VERIFIER_LENGTH
+        )
+        code_verifier = self.strategy.random_string(code_verifier_len)
+        self.strategy.session_set(name, code_verifier)
+        return code_verifier
```
Comment (andrey-canon) [inline eox_core/social_tpa_backends.py:227]:
```suggestion
        self.strategy.session_set(name, code_verifier)
   
        return code_verifier
```

File: eox_core/social_tpa_backends.py [hunk 6]
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
...
+
+    def get_code_verifier(self):
+        name = self.name + "_code_verifier"
```
Comment (andrey-canon) [inline eox_core/social_tpa_backends.py:230]:
```suggestion
        name = f"{self.name}_code_verifier"
```

File: eox_core/social_tpa_backends.py [hunk 7]
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
...
+        code_verifier = self.strategy.session_get(name)
+        return code_verifier
```
Comment (andrey-canon) [inline eox_core/social_tpa_backends.py:232]:
```suggestion
        code_verifier = self.strategy.session_get(name)
   
        return code_verifier
```

File: eox_core/social_tpa_backends.py [hunk 8]
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
...
+
+    def generate_code_challenge(self, code_verifier, challenge_method):
+        method = challenge_method.lower()
+        if method == "s256":
+            hashed = hashlib.sha256(code_verifier.encode()).digest()
+            encoded = base64.urlsafe_b64encode(hashed)
+            code_challenge = encoded.decode().replace("=", "")  # remove padding
+            return code_challenge
+        elif method == "plain":
```
Comment (andrey-canon) [inline eox_core/social_tpa_backends.py:241]:
```suggestion
            return code_challenge

        if method == "plain":
```

File: eox_core/social_tpa_backends.py [hunk 9]
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
...
+            return code_verifier
+        else:
+            raise AuthException("Unsupported code challenge method.")
```
Comment (andrey-canon) [inline eox_core/social_tpa_backends.py:244]:
```suggestion

        raise AuthException("Unsupported code challenge method.")
```

File: eox_core/social_tpa_backends.py [hunk 10]
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
...
+    Base class for providers using OAuth2 with Proof Key for Code Exchange (PKCE).
+    OAuth2 details at:
+        https://datatracker.ietf.org/doc/html/rfc6749
+    PKCE details at:
+        https://datatracker.ietf.org/doc/html/rfc7636
+    """
+
+    PKCE_DEFAULT_CODE_CHALLENGE_METHOD = "s256"
+    PKCE_DEFAULT_CODE_VERIFIER_LENGTH = 32
+    USE_PKCE = True
```
Comment (andrey-canon) [inline eox_core/social_tpa_backends.py:218]: I don't understand why is this necessary, I mean the obvious answer is because you can disable that behavior on  sub-classes, but if I don't want that behavior on sub-classes I could inherit from  BaseOAuth2 instead of this one

File: eox_core/social_tpa_backends.py [hunk 11]
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
...
+
+        if self.USE_PKCE:
+            code_challenge_method = self.setting(
+                "PKCE_CODE_CHALLENGE_METHOD",
+                default=self.PKCE_DEFAULT_CODE_CHALLENGE_METHOD,
+            )
+            code_verifier = self.create_code_verifier()
+            code_challenge = self.generate_code_challenge(
+                code_verifier, code_challenge_method
+            )
```
Comment (andrey-canon) [inline eox_core/social_tpa_backends.py:257]:
```suggestion
            code_challenge = self.generate_code_challenge(
                code_verifier,
                code_challenge_method,
            )
```

File: eox_core/social_tpa_backends.py [hunk 12]
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
...
+            params["code_challenge_method"] = code_challenge_method
+            params["code_challenge"] = code_challenge
+        return params
```
Comment (andrey-canon) [inline eox_core/social_tpa_backends.py:260]:
```suggestion
            params["code_challenge"] = code_challenge

        return params
```

File: eox_core/social_tpa_backends.py [hunk 13]
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
...
+        code_verifier = self.strategy.session_get(name)
+        return code_verifier
+
+    def generate_code_challenge(self, code_verifier, challenge_method):
+        method = challenge_method.lower()
+        if method == "s256":
+            hashed = hashlib.sha256(code_verifier.encode()).digest()
+            encoded = base64.urlsafe_b64encode(hashed)
+            code_challenge = encoded.decode().replace("=", "")  # remove padding
+            return code_challenge
```
Comment (andrey-canon) [inline eox_core/social_tpa_backends.py:240]:
```suggestion
            return encoded.decode().replace("=", "")  # remove padding
```

File: eox_core/social_tpa_backends.py [hunk 14]
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
...
+    """
+
+    PKCE_DEFAULT_CODE_CHALLENGE_METHOD = "s256"
+    PKCE_DEFAULT_CODE_VERIFIER_LENGTH = 32
+    USE_PKCE = True
+
+    def create_code_verifier(self):
+        name = self.name + "_code_verifier"
+        code_verifier_len = self.setting(
+            "PKCE_CODE_VERIFIER_LENGTH", default=self.PKCE_DEFAULT_CODE_VERIFIER_LENGTH
```
Comment (andrey-canon) [inline eox_core/social_tpa_backends.py:223]:
```suggestion
            "PKCE_CODE_VERIFIER_LENGTH",
            default=self.PKCE_DEFAULT_CODE_VERIFIER_LENGTH,
```

**Review by andrey-canon — COMMENTED** [review]
LGTM,  I just left some comments related with the code's style that I would have implemented , however I know this is a copy of the social-core pr  so feel free to ignore these I just wanted to point them out

**Comment by johanv26** [comment]
@andrey-canon I updated to the version is approved in social_core. But taking advantage of that PR is not merged yet, I added some of your review to that PR.
https://github.com/python-social-auth/social-core/pull/856#pullrequestreview-1742515055
Maybe they pay attention and would be included in the social_core version.
